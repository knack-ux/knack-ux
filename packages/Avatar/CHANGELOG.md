# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.5.4](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.5.3...@knack-ux/avatar@0.5.4) (2020-02-29)

**Note:** Version bump only for package @knack-ux/avatar





## [0.5.3](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.5.2...@knack-ux/avatar@0.5.3) (2020-02-25)


### Bug Fixes

* **avatar:** change default variant to info ([886db17](https://github.com/knack-ux/knack-ux/commit/886db1763a6831ff3aec55189b1c906222972a5b))





## [0.5.2](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.5.1...@knack-ux/avatar@0.5.2) (2020-02-12)

**Note:** Version bump only for package @knack-ux/avatar





## [0.5.1](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.5.0...@knack-ux/avatar@0.5.1) (2020-01-23)

**Note:** Version bump only for package @knack-ux/avatar





# [0.5.0](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.4.1...@knack-ux/avatar@0.5.0) (2020-01-23)


### Features

* **avatar:** add icon support and aria-label properties ([32d96c6](https://github.com/knack-ux/knack-ux/commit/32d96c6))





## [0.4.1](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.4.0...@knack-ux/avatar@0.4.1) (2020-01-20)

**Note:** Version bump only for package @knack-ux/avatar





# [0.4.0](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.3.3...@knack-ux/avatar@0.4.0) (2020-01-17)


### Features

* **update:** snapshots and dep ([d0f22b1](https://github.com/knack-ux/knack-ux/commit/d0f22b1))





## [0.3.3](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.3.2...@knack-ux/avatar@0.3.3) (2019-12-21)

**Note:** Version bump only for package @knack-ux/avatar





## [0.3.2](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.3.1...@knack-ux/avatar@0.3.2) (2019-12-21)

**Note:** Version bump only for package @knack-ux/avatar





## [0.3.1](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.3.0...@knack-ux/avatar@0.3.1) (2019-12-21)

**Note:** Version bump only for package @knack-ux/avatar





# [0.3.0](https://github.com/knack-ux/knack-ux/compare/@knack-ux/avatar@0.2.0...@knack-ux/avatar@0.3.0) (2019-12-16)


### Features

* **avatar:** create Avatar component ([33322f9](https://github.com/knack-ux/knack-ux/commit/33322f9))





# 0.2.0 (2019-12-14)


### Features

* **icon, avatar:** create components ([0901f92](https://github.com/knack-ux/knack-ux/commit/0901f92))
