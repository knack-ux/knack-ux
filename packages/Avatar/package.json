{
  "name": "@knack-ux/avatar",
  "version": "0.5.0",
  "description": "Avatar component for the Knack UI Project",
  "main": "lib/index.js",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/knack-ux/knack-ux.git",
    "directory": "@knack-ux/avatar"
  },
  "files": [
    "lib/"
  ],
  "author": "Kristiyan Serafimov (chrispcode)",
  "license": "MIT",
  "scripts": {
    "start": "webpack --watch --config=../../webpack/webpack.dev.js",
    "build": "webpack --config=../../webpack/webpack.prod.js",
    "build:doc": "webpack --config=../../webpack/webpack.dev.js",
    "clean": "rm -rf lib"
  },
  "publishConfig": {
    "access": "public"
  },
  "peerDependencies": {
    "react": "^16.8.4",
    "styled-components": "^4.1.3",
    "styled-system": "^5.1.2"
  },
  "dependencies": {
    "@knack-ux/icon": "^0.2.4",
    "@knack-ux/theme": "^0.7.2"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.0.0",
    "@testing-library/react": "^9.4.0"
  },
  "gitHead": "259d59c20c130d889bb07835c16f1b8bda3dff43"
}
