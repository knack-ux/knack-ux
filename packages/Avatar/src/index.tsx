
import React from 'react';
import styled from 'styled-components';
import {
  layout, space, color,
  LayoutProps, SpaceProps, ColorProps
} from 'styled-system';

export function generateInitials(name: string) {
  const nameParts = name.split(' ');
  let initials = nameParts[0].substring(0, 1).toUpperCase();

  if (nameParts.length > 1) {
    initials += nameParts[nameParts.length - 1].substring(0, 1).toUpperCase();
  }
  return initials;
}

type ReactDiv = Omit<React.HTMLAttributes<HTMLDivElement>, 'color'>;

export interface Props extends ReactDiv, LayoutProps, SpaceProps, ColorProps {
  name: string
  src?: string
}

export function Avatar({
  name,
  src,
  ...props
}: Props) {
  function renderImage() {
    return src && (
      <img src={src} alt="User Avatar" />
    );
  }

  function renderInitial() {
    return !src && (
      generateInitials(name)
    );
  }

  return (
    <AvatarBase
      color="white"
      backgroundColor="default"
      {...props}
    >
      {renderInitial()}
      {renderImage()}
    </AvatarBase>
  );
}

const AvatarBase: React.ComponentType<Omit<Props, 'name'>> = styled.div`
  display: inline-flex;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  
  font-size: 18px;
  font-family: "Cabin";
  font-weight: 600;

  overflow: hidden;

  ${layout}
  ${space}
  ${color}
`;

export default Avatar;
