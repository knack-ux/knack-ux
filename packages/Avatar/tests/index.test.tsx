import React from 'react';
import { render } from '@testing-library/react';

import Avatar from '../src';

describe('<Avatar /> component', () => {
  it('match snapshot on 40px size (initials)', () => {
    const { container } = render(
      <Avatar name="Jon Snow" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('match snapshot on 24px size (initials)', () => {
    const { container } = render(
      <Avatar name="Jon Snow" size={24} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('match snapshot on 40px size (image src)', () => {
    const { container } = render(
      <Avatar
        name="Jon Snow"
        src="https://upload.wikimedia.org/wikipedia/commons/d/d0/Nikola-tesla-3909844.jpg"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
