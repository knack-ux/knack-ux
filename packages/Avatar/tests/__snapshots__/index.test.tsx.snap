// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> component match snapshot on 24px size (initials) 1`] = `
<div
  aria-label="Jon Snow"
  class="sc-AykKC cLAIFd"
  color="white"
  font-size="11"
  size="24"
>
  J
</div>
`;

exports[`<Avatar /> component match snapshot on 40px size (image src) 1`] = `
<div
  aria-label="Jon Snow"
  class="sc-AykKC iLXZeW"
  color="white"
  font-size="16"
  size="40"
>
  <img
    alt="Jon Snow"
    src="https://upload.wikimedia.org/wikipedia/commons/d/d0/Nikola-tesla-3909844.jpg"
  />
</div>
`;

exports[`<Avatar /> component match snapshot on 40px size (initials) 1`] = `
<div
  aria-label="Jon Snow"
  class="sc-AykKC iLXZeW"
  color="white"
  font-size="16"
  size="40"
>
  JS
</div>
`;
